go-controller/pkg/ovn/policy_test.go

```diff
@@ -1522,6 +1524,100 @@ var _ = ginkgo.Describe("OVN NetworkPolicy Operations", func() {
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
 		})
 
+		ginkgo.It("reconciles a completed and deleted pod whose IP has been assigned to a running pod", func() {
+			watch.DefaultChanSize = 300
+			app.Action = func(ctx *cli.Context) error {
+				namespace1 := *newNamespace(namespaceName1)
+				nodeName := "node1"
+
+				// Use simple allow-same-namespace network policy
+				networkPolicy := newNetworkPolicy("networkpolicy1", namespace1.Name,
+					metav1.LabelSelector{},
+					[]knet.NetworkPolicyIngressRule{{
+						From: []knet.NetworkPolicyPeer{{
+							PodSelector: &metav1.LabelSelector{},
+						}},
+					}},
+					[]knet.NetworkPolicyEgressRule{})
+
+				fakeOvn.startWithDBSetup(initialDB,
+					&v1.NamespaceList{
+						Items: []v1.Namespace{
+							namespace1,
+						},
+					},
+					&knet.NetworkPolicyList{
+						Items: []knet.NetworkPolicy{
+							*networkPolicy,
+						},
+					},
+				)
+
+				err := fakeOvn.controller.lsManager.AddSwitch(
+					nodeName,
+					getLogicalSwitchUUID(fakeOvn.controller.nbClient, nodeName),
+					[]*net.IPNet{ovntest.MustParseIPNet("10.128.1.0/29")})
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+				fakeOvn.controller.WatchNamespaces()
+				fakeOvn.controller.WatchPods()
+				fakeOvn.controller.WatchNetworkPolicy()
+
+				// Run job like pods to use all the IPs in the pool  10.128.1.0/29 (10.128.1.3 -> 10.128.1.6)
+				completedPods := make([]*kapi.Pod, 0)
+				for i := 3; i < 7; i++ {
```

Comment:
do you really need to create 4 pods here? can't you just create one pod, then mark it completed, then create a new pod with the same ip, and delete the first one?

Reply:
Unfortunately, yes, we need to exhaust the node's IP pool. Despite what  IP we pass to the `newPod(...)` function, the IP allocator decides the pod IP address. Maybe we get luckier by setting annotations on the created pod:

https://github.com/ovn-org/ovn-kubernetes/blob/cd78ae1af4657d38bdc41003a8737aa958d62b9d/go-controller/pkg/ovn/base_network_controller_pods.go#L580

Reply:
@zeeke take a look at https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/pods_test.go#L437

Reply:
@trozet Much better, thanks!